src/gui/toolbarMenubar/ToolPageSpinner.cpp

```diff
@@ -35,24 +62,57 @@ auto ToolPageSpinner::getNewToolIcon() -> GtkWidget* {
 }
 
 auto ToolPageSpinner::newItem() -> GtkToolItem* {
-    GtkToolItem* it = gtk_tool_item_new();
-
-    GtkWidget* hbox = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 1);
+    if (this->pageSpinner->hasWidget()) {
+        this->pageSpinner->removeWidget();
+    }
+    GtkWidget* spinner = gtk_spin_button_new_with_range(0, 1, 1);
+    this->pageSpinner->setWidget(spinner);
+    gtk_orientable_set_orientation(reinterpret_cast<GtkOrientable*>(spinner), orientation);
 
     GtkWidget* pageLabel = gtk_label_new(_("Page"));
-    GtkWidget* spinner = this->pageSpinner->getWidget();
+    this->lbPageNo = gtk_label_new("");
 
-    gtk_widget_set_valign(pageLabel, GTK_ALIGN_BASELINE);
-    gtk_box_pack_start(GTK_BOX(hbox), pageLabel, false, false, 7);
+    if (orientation == GTK_ORIENTATION_HORIZONTAL) {
+        this->lbVerticalPdfPage = nullptr;
+        gtk_widget_set_valign(pageLabel, GTK_ALIGN_BASELINE);
+        gtk_widget_set_valign(spinner, GTK_ALIGN_BASELINE);
+        gtk_widget_set_valign(this->lbPageNo, GTK_ALIGN_BASELINE);
+    } else {
+        this->lbVerticalPdfPage = gtk_label_new("");
+        gtk_widget_set_halign(pageLabel, GTK_ALIGN_BASELINE);
+        gtk_widget_set_halign(spinner, GTK_ALIGN_CENTER);
+        gtk_widget_set_halign(this->lbPageNo, GTK_ALIGN_BASELINE);
+        gtk_widget_set_halign(lbVerticalPdfPage, GTK_ALIGN_BASELINE);
+    }
 
-    gtk_widget_set_valign(spinner, GTK_ALIGN_BASELINE);
-    gtk_box_pack_start(GTK_BOX(hbox), spinner, false, false, 0);
+    updateLabels();
 
-    this->lbPageNo = gtk_label_new("");
-    gtk_widget_set_valign(this->lbPageNo, GTK_ALIGN_BASELINE);
-    gtk_box_pack_start(GTK_BOX(hbox), this->lbPageNo, false, false, 7);
+    box = gtk_box_new(orientation, 1);
+    gtk_box_pack_start(GTK_BOX(box), pageLabel, false, false, 7);
+    gtk_box_pack_start(GTK_BOX(box), spinner, false, false, 0);
+    gtk_box_pack_start(GTK_BOX(box), this->lbPageNo, false, false, 7);
 
-    gtk_container_add(GTK_CONTAINER(it), hbox);
+    GtkToolItem* it = gtk_tool_item_new();
+    gtk_container_add(GTK_CONTAINER(it), box);
 
     return it;
 }
+
+auto ToolPageSpinner::createItem(bool horizontal) -> GtkToolItem* {
+    this->orientation = horizontal ? GTK_ORIENTATION_HORIZONTAL : GTK_ORIENTATION_VERTICAL;
+
+    this->item = createTmpItem(horizontal);
+    g_object_ref(this->item);
+
+    if (GTK_IS_TOOL_BUTTON(this->item) || GTK_IS_TOGGLE_TOOL_BUTTON(this->item)) {
+        g_signal_connect(this->item, "clicked", G_CALLBACK(&toolButtonCallback), this);
+    }
+
+    return this->item;
+}
+
+auto ToolPageSpinner::createTmpItem(bool horizontal) -> GtkToolItem* {
```

Comment:
As far as I can see this is only used once in the whole code. It would make sense to not have this in its own method then.

Reply:
I removed some obsolete code in my latest commit. I think that we need this method because it overrides the Method of AbstractToolItem. When I remove it or move the code to the createItem Method I get GTK assertions because the g_object_ref is missing.

Reply:
Okay that might be the case. Then the override flag is missing in the header. It would have made this restriction obvious.

Reply:
I've added it to all overridden methods in the ToolPageSpinner header.